README.md
# Autocrackeo
Programa en python que automatiza el uso del hashcat para crackear contraseñas

Específicamente pensado para el caso de no tener excesivos recursos y no poder afrontar grandes ataques de fuerza bruta. En este caso es interesante tener a mano un amplio conjunto de recursos como diccionarios, máscaras y reglas que maximicen el resultado de hashes crackeados en el menor tiempo posible.

La idea es tener varios archivos como fast.json, basic.json, full.json, etc. con ataques de hashcat predefinidos que se ajusten a la velocidad/eficiencia que se requiera en cada momento. Al ejecutar el autocrackeo especificando uno de estos archivos, el programa ejecutará secuencialmente (sin necesidad de supervisión) los ataques de hashcat con los diccionarios, reglas y máscaras definidos en él.

Para ejecutar un archivo de ataques: (ejemplo para lanzar fast.json)
> python3 autocrackeo.py -m 1000 -i docs\test_files\ntlm.hash  -w docs\test_files\custom.dic -o docs\test_files\results --feedback -a fast --verbose

Para ejecutar secuencialmente diferentes archivos de ataques: (ejemplo para lanzar los archivos de ataques listados en all.json)
> python3 autocrackeo.py -m 1000 -i docs\test_files\ntlm.hash  -w docs\test_files\custom.dic -o docs\test_files\results --feedback -a all --verbose


Nueva GUI para generar y ejecutar el comando de forma más visual: `python3 autocrackeo-gui.py`
![Autocrackeo GUI](docs/autocrackeo-gui.PNG)

Ejecución de autocrackeo:
![Autocrackeo](docs/autocrackeo.PNG)

Archivos de resultados generados:
![Resultados](docs/results.PNG)

## Requisitos
Probado en windows/linux con python 3
* [python3](https://www.python.org/downloads/)
* [hashcat](https://github.com/hashcat/hashcat)
	* Hay que tener en cuenta que dependiendo de la versión del hashcat (en windows y a partir de la 3.6) te fuerza a lanzar el hashcat únicamente desde su propio directorio.

## Configuración inicial

Instalar módulos requeridos:
```
pip3 install -r requirements.txt
```

Descargar recursos externos semi-automáticamente: (en proceso de mejora)
* Ejecutar el script de configuración donde están los enlaces a recursos externos que son muy interesantes (diccionarios, reglas, máscaras...)
```
python3 setup.py
```

## Manual de usuario

### Por dónde empezar...
Especifica el path al ejecutable de hashcat y las opciones de rendimiento en el archivo de configuración del equipo "src\HOST_CONFIG.json", por ejemplo:
* executable: hashcat64.exe (Windows) y hashcat (Linux) (en linux o  versiones anteriores de hashcat en windows puedes especificar la ruta absoluta)
* resources: Opciones de hashcat para funcionar dependiendo del equipo en el que se ejecuta: -D 1 -w 3 --force ...

Ejecuta el programa: 
* El programa muestra por pantalla la ejecución del hashcat, por lo que se puede interactuar con él:
	* **s**: mostrar estado (status) / también sirve el "Enter"
	* **p**: pausar (pause)
	* **r**: reanudar (resume)
	* **b**: saltarse fases de un mismo ataque (bypass)
	* **q**: saltarse un ataque completo (quit)
* Ctrl+C para cancelar todos los ataques  sobre un archivo de hashes


### Personalización
1. Modifica el archivo de configuración src\HOST-CONFIG.json para tu entorno y necesidades.
2. Introduce más archivos de diccionarios, reglas y máscaras en los directorios wordlists/, rules/ y masks/ respectivamente.
3. Modifica los archivos de configuración de ataques a tu gusto.
	* Lista los archivos de diccionarios, reglas y máscaras que quieras utilizar en este caso.
	* Define los modos ataque que quieras lanzar con sus parámetros correspondientes.

Nota: El tiempo que tarde en realizar un archivo de ataques dependerá de muchos factores como el número de hashes, el tipo de hash, el tamaño de los diccionarios, de las máscaras, la capacidad del equipo, etc.

### Ayuda
Para ver todas las opciones:
```python
python3 autocrackeo.py -h
```

### Ejemplos
Ejemplo más sencillo: lanza comandos de hashcat sobre el archivo de hashes de entrada  
Tenéis unos archivos de prueba en la siguiente ruta autocrackeo\docs\test_files: ntlm.hash y custom.dic
```console
python3 %PATH_AUTOCRACKEO%/autocrackeo.py -i %PATH_PROYECTO%\ntlm.hash -m 1000 -a all
```

Un archivo: pasando un diccionario de entrada personalizado y diciéndole donde dejar los resultados
```console
python3 %PATH_AUTOCRACKEO%/autocrackeo.py -m 1000 -i %PATH_PROYECTO%\ntlm.hash -w %PATH_PROYECTO%\custom_wordlist.dic -o %PATH_PROYECTO%\results_dir -e="--username" -a quick_test.json
```

Varios archivos secuencialmente: para más información ver explicación en la carpeta "docs/Ejemplos"
```console
python3 %PATH_AUTOCRACKEO%/autocrackeo.py -I %PATH_PROYECTO%\hash_files_list.json -w %PATH_PROYECTO%\custom_wordlist.dic -o %PATH_PROYECTO%\results -e="--username" -a quick_test.json 
```

Nota: en algunas versiones de hashcat te fuerza a ejecutar el hashcat desde el propio directorio del ejecutable, así que modificar las rutas a los archivos para ese caso concreto. Por ejemplo:
```console
cd \tools\hashes\hashcat-5.1.0\
python3 %PATH_AUTOCRACKEO%\autocrackeo.py -m ntlm  -i %PATH_PROYECTO%\ntlm.hash  -w %PATH_PROYECTO%\custom_wordlist.dic -o %PATH_PROYECTO%\results -a quick_test.json
```

Se recomienda el uso de las siguientes opciones:
* --feedback: cogerá las contraseñas del potfile y las volcará si no existen aún al archivo de diccionario personalizado (-w)
* --verbose: para ver más información sobre los comandos de hashcat que se van ejecutando uno tras otro
* -e="--username": si el archivo de hashes está con el formato "usuario:hash" se puede añadir la oopción "--username" en hashcat para que los resultados del potfile salgan con formato "usuario:contraseña" en lugar de únicamente la "contraseña".


## Recomendaciones de uso

### Recursos de proyecto

Se recomienda tener bajo un mismo directorio todos los recursos específicos de un mismo proyecto.

Hashes:
* almacenar todos los archivos con hashes que se recopilen para ese proyecto

Diccionarios:
* Diccionarios con palabras/frases específicas del proyecto (datos de cliente, servicios, usuarios...)


### Recursos de autocrackeo
Los enlaces a todos ellos para descargarlos de una vez se encuentran en el archivo setup.py:

Diccionarios: wordlists/
* cracked.dic: añadir aquí las contraseñas que vayas crackeando en cada proyecto para que te sirvan para los posteriores
* super.dic: recolección de palabras genéricas, lugares, colores, nombres, fechas... susceptibles de aparecer en contraseñas
* añadir más diccionarios:
	* rockyou: http://downloads.skullsecurity.org/passwords/rockyou.txt.bz2
	* hashkiller: https://hashkiller.co.uk/downloads.aspx
	* colecciones de usuarios, contraseñas, urls, patrones...: https://github.com/danielmiessler/SecLists
	* palabras típicas: https://raw.githubusercontent.com/first20hours/google-10000-english/master/google-10000-english.txt

Reglas: rules/
* fast.rule, basic.rule, full.rule: creadas específicamente para esta herramienta según en nivel de complejidad deseado
* añadir más reglas:
	* best64.rule, d3ad0ne.rule y T0XIC.rule: https://github.com/hashcat/hashcat/tree/master/rules
	* hob064.rule: https://github.com/praetorian-inc/Hob0Rules
	* OneRuleToRuleThemAll.rule: https://github.com/NotSoSecure/password_cracking_rules
	* toggles-lm-ntlm.rule: https://github.com/trustedsec/hate_crack/blob/master/rules/toggles-lm-ntlm.rule
	* haku34K.rule: https://raw.githubusercontent.com/kaonashi-passwords/Kaonashi/master/rules/haku34K.rule,
	* kamaji34K.rule: https://raw.githubusercontent.com/kaonashi-passwords/Kaonashi/master/rules/kamaji34K.rule,
	* yubaba64.rule: https://raw.githubusercontent.com/kaonashi-passwords/Kaonashi/master/rules/yubaba64.rule
		

Máscaras: masks/
* fast_hybrid.hcmask, basic.hcmask, basic_hybrid.hcmask, full.hcmask, full_hybrid.hcmask: creadas específicamente para esta herramienta según en nivel de complejidad deseado
* añadir más máscaras:
	* 2015-Top40-Time-Sort.hcmask: https://blog.netspi.com/netspis-top-password-masks-for-2015/
	* pathwell.hcmask: https://github.com/trustedsec/hate_crack/blob/master/masks/pathwell.hcmask
	* kaonashi.hcmask: https://raw.githubusercontent.com/kaonashi-passwords/Kaonashi/master/masks/kaonashi.hcmask"
	* rockyou-1-60.hcmask: https://raw.githubusercontent.com/hashcat/hashcat/master/masks/rockyou-1-60.hcmask


Configuraciones de ataques automáticos:
* quick_test.json, fast.json, basic.json, full.json: ataques predefinidos de menos a más complejidad
	* quick_test.json: prueba una vez los diccionarios custom, cracked y super
	* fast.json: configuración para una ejecución rápida con pocos intentos.
	* basic.json: configuración para una ejecución básica, que en minutos/horas sea capaz de probar las contraseñas/reglas/máscaras más frecuentes.
	* full.json: configuración para una ejecución más completa, puede tardar semanas, meses, años incluso. Además de probar las opciones básicas, añadirá reglas combinadas con diccionarios grandes, máscaras que abarquen más caracteres de fuerza bruta, etc.
* all_wordlists_all_rules.json: todas las combinaciones de cada diccionario con cada regla
* one_word_per_hash.json: generar un diccionario sólo con los nombres de usuario y probar únicamente usuario=contraseña
	* puede resultar útil teniendo hashes de formato de hashes complejos (que tardan mucho en cada prueba)
	* el número de líneas del hashlist y wordlist deben coincidir
* lm.json: prueba todas las combinaciones posibles de hashes en formato LM (1-8 caracteres)
* ntlm_from_lm.json: partiendo de un wordlist con contraseñas sacadas en LM (case insensitive) prueba todas las combinaciones de mayúsculas y minúsculas para sacar las contraseñas NTLM (case sensitive)


## Organización
El proyecto está dividido por directorios de distintos recursos. Aunque en ningún momento se restringe completamente a seguir esta estructura. Los directorios de hashes, diccionarios, reglas, máscaras y resultados, se pueden especificar como parámetros de entrada o en la configuración json, por lo que se pueden modificar o unificar perfectamente.

* En el directorio **wordlists/** se almacenaran los diccionarios que contengan palabras o frases susceptibles a aparecer en contraseñas.
* En el directorio **rules/** se almacenaran los archivos con reglas (modificaciones sobre las palabras del diccionario)
* En el directorio **masks/** se almacenaran los archivos con máscaras (patrones de fuerza bruta)
* En el directorio **config/** se almacenarán los archivos json que contienen la configuración y los ataques a ejecutar...
* En el directorio **src/** está el código fuente del programa.
* En el directorio **docs/** se almacenarán archivos de interés relacionados con el proyecto
	* changelog.md: aquí se irán indicando los cambios realizados y las nuevas ideas por desarrollar
	* Tutorial de crackeo (para hashcat).md: aquí he reunido los conceptos y ejemplos del hashcat que me han sido de utilidad para realizar este proyecto.
	* Archivo de hashes y diccionario para pruebas.

## Autora
* **Eneritz Azqueta** → Auditora en **S21sec**

docs/changelog.md

# Cambios

* 06/06/2018:
	* compatibilidad windows/linux

* 16/08/2018:
	* cambiar la entrada de parámetros del teclado al script con opción de ayuda
	* reestructurar el proyecto para poder ampliar el alcance del mismo
		* Archivos de configuración de la ejecución: separar toda la configuración modificable por el usuario en un único archivo JSON
		* Autocrackeo.py: script principal que interpreta la configuración y llama a los métodos de ataque definidos
		* Clase Attack: parsea los datos de ataque y llama a los correspondientes métodos de la clase hashcat
		* Clase Hashcat: específica de hashcat y sus parámetros para facilitar las llamadas a la herramienta
	* crear varios modos de funcionamiento en base a la configuración (json): rápido, básico, completo...
	* analizar, mostrar y almacenar los resultados:
		* outfile=plaintext_passwords.txt: contraseñas en texto plano crackeadas en la ejecución
		* out_file_cracked=hashes_cracked.txt: user:password crackeados
		* out_file_left=hashes_left.txt: user:hash sin crackear
	* añadir opción de ataque one_word_per_hash --> ¿hashlist[i] == hash(wordlist[i])? (línea a línea)

* 25/09/2018:
	* añadir archivos conocidos de hashes, reglas, máscaras y diccionarios.
	* cambiar el orden de los ataques pasando la fuerza bruta al último lugar. En realidad depende del orden en el que se escriba el archivo de configuración.
	* mover la configuración de entorno (executable,resources) a un único fichero "host-config.json", fuera de los config.json.
	* mover la opción de añadir parámetros extra del hashcat como argumentos de entrada --extra-params="--username" o -e="--username", en lugar de en los archivos config.json.
	* actualizar los archivos de ejemplos de configuración y READMEs acorde a los cambios.
	* Añadir búsqueda rápida de tipos de hash, para introducir "NTLM" en lugar de "1000", por ejemplo.

* 06/02/2019: v1.3
	* simplificar la obtención de resultados: mostrar en pantalla sólo comando del hashcat + crackeados al instante, generar archivos con el nombre del hashlist y contenido usuario:contraseña en un directorio concreto al usar -r o --results
	* posibilidad de saltarse la secuencia de ataques y pasar directamente a los resultados con las teclas "Ctrl+c"
	* se ha movido la definición de los directorios base de cada archivo de configuración de ataques, al archivo de configuración general host-conf.json
	* comprobaciones de existencia de ficheros necesarios
	* eliminar posibilidad de elegir ruta a cada directorios, identifica la ruta del archivo de configuración con el parámetro -c, y utiliza como directorio base el directorio anterior.
	* arreglado para que funcione también en windows, que te obliga a ejecutarlo desde el directorio del hashcat.
	* se han añadido las configuraciones all_wordlists_all_rules.json, lm.json y ntlm_from_lm.json
	* ejecutar en bucle un mismo ataque para varios archivos y tipos desde un archivo json con una lista de hash_file, hash_type y extra_params.
	* eliminar innecesario texto con resultados.
	* añadir elección de ruta a /results/ donde dejar el potfile y los resultados de la ejecución
	* cambiar argumento -c para utilizar ruta relativa desde /autocrackeo/config/ con -c test.json o ruta absoluta C:\test\test.json
	* añadir argumento de entrada para elegir el path del archivo custom.dic a utilizar

* 08/02/2019: v1.4
	* añadir pathwell.hcmask mask to all_wordlists_all_rules.json
	* opción -c "all" para que ejecute en secuencialmente cada archivo de configuración definido en el archivo all.json

* 08/02/2019: v1.5
	* cambios en las configuraciones para ajustar los tiempos de ejecución
	* añadidos recursos y ataques del repositorio kaonashi de github

* 07/03/2020: v1.6
	* cambios en el formato de mostrar comentarios por pantalla y colores
	* añadida la opción --feedback  para volcar contraseñas nuevas del potfile al diccionario personalizado (-w)

* 05/04/2020: 1.7
	* pequeños arreglos
	* actualización del script para descargar los recursos externos: setup.py
	* añadido el archivo requirements.txt

* 25/04/2020: v1.8
	* añadida interfaz gráfica con tkinter para lanzar el comando de autocrackeo en un thread aparte
	* añadida la función de logger para guardar en un archivo los comandos que se han ido ejecutando
	* update readme with demo images

* 08/05/2020: v1.9
	* arreglar rutas de autocrackeo si se lanza desde otro directorio
	* quitar del usuario la parte de escribir el .json de los ataques
	* si no incluyes un wordlist con la opción -w, por defecto usa como diccionario primario wordlists\super.dic
	* cambiar nomenclatura de configuración -c por ataque -a, que es menos confuso
	* recuperar contenido original de super.dic que había borrado sin querer
	* gui: en lugar de recalcular el comando y ejecutar, que lea el contenido del text box por si el usuario lo cambmia a mano


# PARA SUBIR V2 

* 02/07/2020: v2.0
	* Añadir recursos corporate masks del repositorio github.com/golem445/Corporate_Masks. Unirlas en un archivo de ataques corporate_masks.json y añadirlo al all.json.
	* Si pulsas 3 veces seguidas en menos de 1 segundo de diferencia Ctrl+c en lugar de saltarse un ataque, para la ejecución de autocrackeo y acaba

# EN PROCESO
* XX/07/2020: v2.0
* acabar setup.py con la parte de descomprimir diccionarios y comprobar que funcione tanto en linux como en windows

# TODO

* GUI: añadir desplegable de ataques (quick_test, fast, basic, custom...)
* separar máscaras de kaonashi entre el basic, full y bruteforce forever
* NTDS analisys
* check how to combine --feedback and --loopback
* search hash from pattern
* add john the ripper for múltiple hybrid attack like ?a?a?w?a?a
* work with machine readable results to update a progress bar
* list of slow/fast hashes to skip some "long/hard" attacks
* more exception handling
* arreglar: Skipping invalid or unsupported rule in file D:\projects\MaiShadow\autocrackeo\rules\OneRuleToRuleThemAll.rule on line 8210: ^o^─à^─é^o^t


autocrackeo.py
# -*- coding: utf-8 -*-
# Author: MaiShadow

"""
 Import modules
"""
try:
	import os, sys, json, argparse, logging
	version = str(sys.version_info[0]) + "." + str(sys.version_info[1])
	if sys.version_info[0] < 3 or sys.version_info[1] < 6:
	    sys.exit("[X] Python version must be at least 3.6, current is " + version)

	import time
	from datetime import datetime, timedelta
	from src.color import Color
	from src.configuration import Configuration
	from src.attacks import Attacks
	from src.hashcat import Hashcat

except Exception as e:
	sys.exit(" [X] " + str(e))

"""
 Manage input parameters
"""
def getArguments():
	parser = argparse.ArgumentParser(description="Automated Hashcat usage tool", epilog='Usage: python3 autocrackeo.py -m 1000 -i docs\\test_files\\ntlm.hash -w docs\\test_files\\custom.dic -o docs\\test_files\\results -a all -e="--username" --feedback --verbose')
	
	# or hashfile + hashtype, or hash files list
	parser.add_argument("-m", type=str, dest="hash_type", help="hashcat's hash type number or its corresponding title, more info here: https://hashcat.net/wiki/doku.php?id=example_hashes")
	group = parser.add_mutually_exclusive_group(required=True)
	group.add_argument("-i", "--hash-file", type=str,  dest="hash_file", help="path to the file with hashes to crack")
	group.add_argument("-I", "--hash-files-list", type=str,  dest="hash_files", help="path to the json file with a list of hash files and types to crack")
	
	# -c  execute attacks
	parser.add_argument("-a", "--attacks", type=str, dest="attacks_file", help="filename to use with specific preconfigured attacks (relative path from: /autocrackeo/attacks/[quick_test].json without the extension")
	
	# -w use custom wordlist from custom path
	parser.add_argument("-w", "--wordlist-custom", type=str, dest="wordlist_custom_file", help="custom primary wordlist to use instead of default /autocrackeo/wordlists/super.dic")

	# extra params to add at the end of the hashcat command
	parser.add_argument("-e", "--extra-params", type=str, dest="extra_params", default="", help="extra params to add in the hashcat command")
	
	# other functionalities
	parser.add_argument("--feedback", action="store_true", dest="feedback", help="dump plaintext passwords from potfile (-o) to custom wordlist (-w)")
	parser.add_argument("-v", "--verbose", action='store_true', dest="verbose", help="show more messages")
	parser.add_argument("--version", action='version', version='%(prog)s 1.8 (25/04/2020)')

	# output directory, default current dir
	parser.add_argument("-o", "--output-dir", type=str, dest="output_dir", default="", help="path to store the results (potfile.pot, cracked.txt, user_pwd.txt")
	return parser.parse_args()

def main(color):
	"""
	 configuration: all the attacks and config data from the user and files
	 attacks: configure attacks from files data
	 hashcat: calls to hashcat individual attacks
	"""

	# print start datetime
	start_date = datetime.now()
	Color.showVerbose("Start date: " + Color.datetime_to_string(start_date))
	Color.showVerbose("Press enter or [s] to see hashcat's status...")
	Color.showVerbose("Press [q]' to skip one hashcat command...")
	Color.showVerbose("Press [Ctrl+c] to skip one attack file...")
	Color.showVerbose("Press [Ctrl+c] x3 times to stop all attacks...")

	# get input arguments
	arguments = getArguments()

	if not arguments.attacks_file:
		Color.showError("Nothing happening here... add [-a attacks_file] to execute attacks", True)

	# ctrl+c signal counter for exiting program instead of bypassing resources
	interruptCounter = 0
	lastInterruptTime = datetime.now()

	try:
		"""
		For every attacks_file in the list (all configs)
		"""
		attacks_file_list = Configuration.getConfigFilesArray(arguments.attacks_file)

		for attacks_file in attacks_file_list:

			"""
			For every hash_file in the list, execute all the defined attacks in the attacks_file
			"""
			hash_files_list = Configuration.getHashFilesArray(arguments.hash_file, arguments.hash_type, arguments.extra_params, arguments.hash_files)

			for hash_file_item in hash_files_list:
				parsing_errors = False

				if hash_file_item:
					if hash_file_item["hash_file"]:
						hash_file = hash_file_item["hash_file"]
					else:
						parsing_errors = True
					if hash_file_item["hash_type"]:
						hash_type = hash_file_item["hash_type"]
					else:
						parsing_errors = True
					if len(hash_file_item) == 3 and hash_file_item["extra_params"]:
						extra_params = hash_file_item["extra_params"]
					else:
						extra_params = ""
				else:
					parsing_errors = True

				if parsing_errors:
					Color.showError("Error in the files/types/extra_param parsing... skipping this file", False)
					break

				# load other scripts
				conf = Configuration(hash_file, hash_type, attacks_file, extra_params, arguments.output_dir, arguments.wordlist_custom_file)
				hashcat = Hashcat(conf.static_values, arguments.verbose, color)
				attacks = Attacks(hashcat)

				
				#set logging file
				log_path = os.path.join(conf.results_dir, "autocrackeo.log")
				color.setFileHandler(log_path) # set log file
				Color.showVerbose("The results (potfile, cracked passwords and logfile) will be written to: " + conf.results_dir + "\n")

				# print important info
				Color.showTitle(Color.datetime_to_string(datetime.now()))
				msg = "Attacks config file:" + attacks_file + ", hash file:" + hash_file + ", hash type:" + hash_type + ", extra params:" + extra_params
				Color.showMessage(msg  + "\n") # show attack file
				color.logThis("[i] " + Color.datetime_to_string(datetime.now()) + ", " + msg) # log attack file

				if attacks_file: # if -c/--config
					"""
					Execute a specific selection of hashcat attacks
					previously defined on the configuration json file
					This will be updated gradually as the efficiency of the attacks are measured
					"""
					try:
						for attack_name in conf.attacks:
							if arguments.verbose: Color.showVerbose("Attack type: " + attack_name.replace("_"," ").title()) # nice print
							if "straight" in attack_name:
								attacks.straight_attacks(attack_name, conf.attacks[attack_name], conf.wordlists, conf.rules)
							elif "combinator" in attack_name:
								attacks.combinator_attacks(attack_name, conf.attacks[attack_name], conf.wordlists)
							elif "brute_force" in attack_name:
								attacks.brute_force_attacks(attack_name, conf.attacks[attack_name], conf.masks)
							elif "hybrid" in attack_name:
								attacks.hybrid_attacks(attack_name, conf.attacks[attack_name], conf.wordlists, conf.masks)
							elif "one_word_per_hash" in attack_name:
								attacks.OneWordPerHashAttacks(attack_name, conf.attacks[attack_name], conf.wordlists)
							else:
								Color.showError("This attack name is not recognized!", False)

							# dump plaintext passwords from potfile to custom wordlist
							if arguments.feedback: hashcat.feedback(arguments.wordlist_custom_file)

					except KeyboardInterrupt as ki:
						"""
						Set a SIGINT signal handler 
						to securely skip all the attacks for this hash_file and attacks_file
						and it continues the loop
						but if it is x3 times clicked with less than 1 second of distance, exit program
						"""
						Color.showError("Ctrl+C {0}: Skipping {1} attacks...".format(interruptCounter+1,attacks_file), False)
						color.logThis("[X] Ctrl+C {0}: Skipping {1} attacks...".format(interruptCounter+1,attacks_file))

						hashcat.save_cracked() # dump results output in cracked file

						interruptTime = datetime.now()
						difference = interruptTime - lastInterruptTime
						if (difference.total_seconds() < 1):
							interruptCounter+=1
							if (interruptCounter > 2):
								ki.message = "ctrl+c x3"
								raise
						else:
							interruptCounter = 0
						
						lastInterruptTime = interruptTime

					except Exception as e:
						Color.showException(e, True)

					hashcat.save_cracked() # ALWAYS DUMP RESULTS: for every config file tried, and every hash file/type

	except KeyboardInterrupt as ki:
		Color.showError("Ctrl+C (x3): Exiting attacks...", False)
		color.logThis("[X] Ctrl+C (x3): Exiting attacks...")
	except Exception as e:
		Color.showError(str(e), False)

	"""
	 Print end of execution
	"""	
	# print end datetime and duration
	Color.showTitle("")
	end_date = datetime.now()
	Color.showVerbose("End date: " + Color.datetime_to_string(end_date))
	duration = end_date - start_date
	Color.showVerbose("Duration: " + Color.timedelta_to_string(duration))

"""
 Call main method
"""
if __name__ == '__main__':
	os.system("") # do nothing but enable command colors on windows cmd
	color = Color()
	Color.showTitle("Autocrackeo")
	try:
		main(color)
	except Exception as e:
		Color.showException(e, True)
	Color.showEnding()